src/Widget.php

```diff
@@ -0,0 +1,182 @@
+<?php
+declare(strict_types=1);
+
+namespace Yiisoft\Widget;
+
+use Psr\EventDispatcher\EventDispatcherInterface;
+use Yiisoft\Widget\Exception\InvalidConfigException;
+use Yiisoft\Widget\Event\AfterRun;
+use Yiisoft\Widget\Event\BeforeRun;
+use Yiisoft\Widget\Factory\WidgetFactory;
+
+/**
+ * Widget is the base class for widgets.
+ */
+class Widget
+{
+    /**
+     * @var EventDispatcherInterface $eventDispatcher
+     */
+    protected EventDispatcherInterface $eventDispatcher;
+
+    /**
+     * The widgets that are currently being rendered (not ended). This property is maintained by {@see begin()} and
+     * {@see end} methods.
+     *
+     * @var array $stack
+     */
+    protected static array $stack;
+
+    /**
+     * @var Widget $widget
+     */
+    protected static Widget $widget;
+
+    public function __construct(EventDispatcherInterface $eventDispatcher)
+    {
+        $this->eventDispatcher = $eventDispatcher;
+    }
+
+    public function init(): void
+    {
+    }
+
+    public function run(): string
+    {
+        return '';
+    }
+
+    /**
+     * Begin the rendering of content.
+     *
+     * @return Widget
+     */
+    public static function begin(): Widget
+    {
+        $widget = WidgetFactory::createWidget(get_called_class());
+
+        static::$stack[] = $widget;
+
+        return $widget;
+    }
+
+    /**
+     * Ends the rendering of content.
+     *
+     * @return Widget
+     */
+    public static function end(): Widget
+    {
+        if (empty(self::$stack)) {
+            throw new InvalidConfigException(
+                'Unexpected ' . static::class . '::end() call. A matching begin() is not found.'
+            );
+        }
+
+        $widget = array_pop(self::$stack);
+
+        if (get_class($widget) !== static::class) {
+            throw new InvalidConfigException('Expecting end() of ' . get_class($widget) . ', found ' . static::class);
+        }
+
+        if ($widget->beforeRun()) {
+            $result = $widget->run();
+            $result = $widget->afterRun($result);
+            echo $result;
```

Comment:
Maybe we should return here? Then we'll need no buffering.

That would make syntax

```php
<?php MyWidget::begin()->name('test) ?>

Hello!

<?= MyWidget::end() ?>
```